tests/devices/qubit/test_apply_operation.py

```diff
@@ -68,6 +84,133 @@ def matrix(self):
     assert qml.math.allclose(new_state, mat @ state)
 
 
+class TestSparseOperation:
+    """Test the sparse matrix application method"""
+
+    ops_to_sparsify = [
+        qml.PauliX(0),
+        qml.CNOT((0, 1)),
+        qml.Toffoli((0, 1, 2)),
+        qml.MultiControlledX(wires=[0, 1, 2, 3, 4], control_values=[1, 1, 1, 1]),
+        qml.GroverOperator(wires=[0, 1, 2]),
+        qml.IsingXX(np.pi / 2, wires=[0, 1]),
+        qml.DoubleExcitation(np.pi / 4, wires=[0, 1, 2, 3]),
+    ]
+
+    def test_sparse_operation_dense_state(self):
+        """Test that apply_operation works with a sparse matrix operation"""
+
+        # Create a random unitary matrix
+        U = unitary_group.rvs(2**3)
+        U = qml.math.asarray(U, like="numpy")
+
+        # Create a random state vector
+        state = np.random.rand(2**3) + 1j * np.random.rand(2**3)
+        state = qml.math.asarray(state, like="numpy").reshape([2] * 3)
+
+        # Apply the operation
+        U_sp = qml.QubitUnitary(csr_matrix(U), wires=range(3))
+        new_state = apply_operation_csr_matrix(U_sp, state)
+        expected_state = state.reshape((1, 8)) @ U.reshape((8, 8)).T
+        expected_state = expected_state.reshape([2] * 3)
+
+        assert qml.math.allclose(new_state, expected_state)
+
+    def test_sparse_operation_sparse_state(self):
+        """Test that apply_operation does not support with a sparse state operation"""
+
+        # Create a random unitary matrix
+        U = unitary_group.rvs(2**3)
+        U = qml.math.asarray(U, like="numpy")
+
+        # Create a random state vector
+        state = np.random.rand(2**3) + 1j * np.random.rand(2**3)
+        state = csr_matrix(state)
+        U_sp = qml.QubitUnitary(csr_matrix(U), wires=range(3))
+
+        # Apply the operation
+        with pytest.raises(
+            TypeError,
+            match="State should not be sparse",
+        ):
+            apply_operation_csr_matrix(U_sp, state)
+
+    @pytest.mark.parametrize("N", range(2, 20, 2))
```

Comment:
I'm not sure we need to check 10 different numbers of wires.  We want to make sure our tests run fast and don't consume a lot of extra resources, especially for particularily expensive tests.  One round of `N=20` should probably be sufficient.

Reply:
https://github.com/PennyLaneAI/pennylane/pull/6883/commits/64be7af5a8888630e503e83b553ee38e52deccd9